File: artifacts/api-server/src/routes/agents.ts
import { Router, type IRouter } from "express";
import { eq } from "drizzle-orm";
import { db, agentsTable, influencesTable, postsTable } from "@workspace/db";
import {
  ListAgentsQueryParams,
  ListAgentsResponse,
  CreateAgentBody,
  GetAgentParams,
  GetAgentResponse,
  UpdateAgentParams,
  UpdateAgentBody,
  UpdateAgentResponse,
  DeleteAgentParams,
  GetAgentNeighborhoodParams,
  GetAgentNeighborhoodResponse,
  CreateInfluenceBody,
} from "@workspace/api-zod";
import { or } from "drizzle-orm";

const router: IRouter = Router();

router.get("/agents", async (req, res): Promise<void> => {
  const query = ListAgentsQueryParams.safeParse(req.query);
  let agents;
  if (query.success && query.data.simulationId) {
    agents = await db
      .select()
      .from(agentsTable)
      .where(eq(agentsTable.simulationId, parseInt(query.data.simulationId, 10)));
  } else {
    agents = await db.select().from(agentsTable).orderBy(agentsTable.createdAt);
  }

  const mapped = agents.map((a) => ({
    ...a,
    beliefState: a.beliefState as { policySupport: number; trustInGovernment: number; economicOutlook: number },
  }));

  res.json(ListAgentsResponse.parse(mapped));
});

router.post("/agents", async (req, res): Promise<void> => {
  const parsed = CreateAgentBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const data = parsed.data;
  const [agent] = await db
    .insert(agentsTable)
    .values({
      name: data.name,
      age: data.age,
      gender: data.gender,
      region: data.region,
      occupation: data.occupation,
      persona: data.persona,
      stance: data.stance,
      influenceScore: data.influenceScore ?? 0.5,
      credibilityScore: data.credibilityScore ?? 0.5,
      beliefState: data.beliefState ?? { policySupport: 0, trustInGovernment: 0.5, economicOutlook: 0.5 },
      confidenceLevel: data.confidenceLevel ?? 0.5,
      activityLevel: data.activityLevel ?? 0.5,
      groupId: data.groupId ?? null,
      simulationId: data.simulationId ?? null,
    })
    .returning();

  const mapped = {
    ...agent,
    beliefState: agent.beliefState as { policySupport: number; trustInGovernment: number; economicOutlook: number },
  };
  res.status(201).json(GetAgentResponse.parse(mapped));
});

router.get("/agents/:id", async (req, res): Promise<void> => {
  const params = GetAgentParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  const [agent] = await db
    .select()
    .from(agentsTable)
    .where(eq(agentsTable.id, params.data.id));

  if (!agent) {
    res.status(404).json({ error: "Agent not found" });
    return;
  }

  const mapped = {
    ...agent,
    beliefState: agent.beliefState as { policySupport: number; trustInGovernment: number; economicOutlook: number },
  };
  res.json(GetAgentResponse.parse(mapped));
});

router.patch("/agents/:id", async (req, res): Promise<void> => {
  const params = UpdateAgentParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  const parsed = UpdateAgentBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const updateData: Record<string, unknown> = {};
  if (parsed.data.name !== undefined) updateData.name = parsed.data.name;
  if (parsed.data.stance !== undefined) updateData.stance = parsed.data.stance;
  if (parsed.data.beliefState !== undefined) updateData.beliefState = parsed.data.beliefState;
  if (parsed.data.confidenceLevel !== undefined) updateData.confidenceLevel = parsed.data.confidenceLevel;
  if (parsed.data.activityLevel !== undefined) updateData.activityLevel = parsed.data.activityLevel;
  if (parsed.data.influenceScore !== undefined) updateData.influenceScore = parsed.data.influenceScore;
  if (parsed.data.credibilityScore !== undefined) updateData.credibilityScore = parsed.data.credibilityScore;

  const [agent] = await db
    .update(agentsTable)
    .set(updateData)
    .where(eq(agentsTable.id, params.data.id))
    .returning();

  if (!agent) {
    res.status(404).json({ error: "Agent not found" });
    return;
  }

  const mapped = {
    ...agent,
    beliefState: agent.beliefState as { policySupport: number; trustInGovernment: number; economicOutlook: number },
  };
  res.json(UpdateAgentResponse.parse(mapped));
});

router.delete("/agents/:id", async (req, res): Promise<void> => {
  const params = DeleteAgentParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  const [agent] = await db
    .delete(agentsTable)
    .where(eq(agentsTable.id, params.data.id))
    .returning();

  if (!agent) {
    res.status(404).json({ error: "Agent not found" });
    return;
  }

  res.sendStatus(204);
});

router.get("/agents/:id/neighborhood", async (req, res): Promise<void> => {
  const params = GetAgentNeighborhoodParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  const [agent] = await db
    .select()
    .from(agentsTable)
    .where(eq(agentsTable.id, params.data.id));

  if (!agent) {
    res.status(404).json({ error: "Agent not found" });
    return;
  }

  const outgoing = await db
    .select()
    .from(influencesTable)
    .where(eq(influencesTable.sourceAgentId, params.data.id));

  const incoming = await db
    .select()
    .from(influencesTable)
    .where(eq(influencesTable.targetAgentId, params.data.id));

  const neighborIds = new Set<number>();
  const connections: Array<{ agent: typeof agent; influenceWeight: number; direction: string }> = [];

  for (const inf of outgoing) {
    neighborIds.add(inf.targetAgentId);
  }
  for (const inf of incoming) {
    neighborIds.add(inf.sourceAgentId);
  }

  for (const nId of neighborIds) {
    const [neighbor] = await db
      .select()
      .from(agentsTable)
      .where(eq(agentsTable.id, nId));

    if (neighbor) {
      const outInf = outgoing.find((i) => i.targetAgentId === nId);
      const inInf = incoming.find((i) => i.sourceAgentId === nId);
      connections.push({
        agent: {
          ...neighbor,
          beliefState: neighbor.beliefState as { policySupport: number; trustInGovernment: number; economicOutlook: number },
        },
        influenceWeight: outInf?.weight || inInf?.weight || 0,
        direction: outInf ? "outgoing" : "incoming",
      });
    }
  }

  const posts = await db
    .select()
    .from(postsTable)
    .where(eq(postsTable.agentId, params.data.id))
    .limit(20);

  const agentMapped = {
    ...agent,
    beliefState: agent.beliefState as { policySupport: number; trustInGovernment: number; economicOutlook: number },
  };

  const postsWithNames = posts.map((p) => ({ ...p, agentName: agent.name }));

  res.json(
    GetAgentNeighborhoodResponse.parse({
      agent: agentMapped,
      connections,
      posts: postsWithNames,
    })
  );
});

router.post("/influences", async (req, res): Promise<void> => {
  const parsed = CreateInfluenceBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const [influence] = await db
    .insert(influencesTable)
    .values({
      sourceAgentId: parsed.data.sourceAgentId,
      targetAgentId: parsed.data.targetAgentId,
      weight: parsed.data.weight,
    })
    .returning();

  res.status(201).json(influence);
});

export default router;

File: artifacts/api-server-py/app/routers/agents.py
import asyncio

from fastapi import APIRouter, HTTPException, Query, Response

from app.db import pool
from app.serialize import agent_row, influence_row, post_row
from app.services import neo4j_service

router = APIRouter()


@router.get("/agents")
async def list_agents(simulationId: str | None = Query(None)) -> list[dict]:
    p = pool()
    async with p.acquire() as conn:
        if simulationId:
            sid = int(simulationId)
            rows = await conn.fetch(
                "SELECT * FROM agents WHERE simulation_id = $1", sid
            )
        else:
            rows = await conn.fetch("SELECT * FROM agents ORDER BY created_at")
    return [agent_row(r) for r in rows]


@router.post("/agents", status_code=201)
async def create_agent(body: dict) -> dict:
    required = (
        "name",
        "age",
        "gender",
        "region",
        "occupation",
        "persona",
        "stance",
    )
    for k in required:
        if k not in body:
            raise HTTPException(status_code=400, detail={"error": f"missing {k}"})

    bs = body.get(
        "beliefState",
        {"policySupport": 0, "trustInGovernment": 0.5, "economicOutlook": 0.5},
    )
    import json

    p = pool()
    async with p.acquire() as conn:
        row = await conn.fetchrow(
            """INSERT INTO agents (
            name, age, gender, region, occupation, persona, stance,
            influence_score, credibility_score, belief_state, confidence_level,
            activity_level, group_id, simulation_id, system_prompt
            ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10::jsonb,$11,$12,$13,$14,$15)
            RETURNING *""",
            body["name"],
            int(body["age"]),
            body["gender"],
            body["region"],
            body["occupation"],
            body["persona"],
            body["stance"],
            float(body.get("influenceScore", 0.5)),
            float(body.get("credibilityScore", 0.5)),
            json.dumps(bs),
            float(body.get("confidenceLevel", 0.5)),
            float(body.get("activityLevel", 0.5)),
            body.get("groupId"),
            body.get("simulationId"),
            body.get("systemPrompt"),
        )
    out = agent_row(row)
    asyncio.create_task(neo4j_service.sync_agent_to_graph(out))
    return out


@router.get("/agents/{id}")
async def get_agent(id: int) -> dict:
    p = pool()
    async with p.acquire() as conn:
        row = await conn.fetchrow("SELECT * FROM agents WHERE id = $1", id)
    if not row:
        raise HTTPException(status_code=404, detail={"error": "Agent not found"})
    return agent_row(row)


@router.patch("/agents/{id}")
async def update_agent(id: int, body: dict) -> dict:
    p = pool()
    async with p.acquire() as conn:
        row = await conn.fetchrow("SELECT * FROM agents WHERE id = $1", id)
        if not row:
            raise HTTPException(status_code=404, detail={"error": "Agent not found"})

        import json

        name = body["name"] if "name" in body else row["name"]
        stance = body["stance"] if "stance" in body else row["stance"]
        conf = float(
            body["confidenceLevel"]
            if "confidenceLevel" in body
            else row["confidence_level"]
        )
        act = float(
            body["activityLevel"]
            if "activityLevel" in body
            else row["activity_level"]
        )
        inf = float(
            body["influenceScore"]
            if "influenceScore" in body
            else row["influence_score"]
        )
        cred = float(
            body["credibilityScore"]
            if "credibilityScore" in body
            else row["credibility_score"]
        )
        b = row["belief_state"]
        if isinstance(b, str):
            b = json.loads(b)
        if "beliefState" in body:
            b = body["beliefState"]

        row = await conn.fetchrow(
            """UPDATE agents SET name=$1, stance=$2, belief_state=$3::jsonb,
            confidence_level=$4, activity_level=$5, influence_score=$6, credibility_score=$7
            WHERE id = $8 RETURNING *""",
            name,
            stance,
            json.dumps(b),
            conf,
            act,
            inf,
            cred,
            id,
        )
    return agent_row(row)


@router.delete("/agents/{id}", status_code=204)
async def delete_agent(id: int) -> Response:
    p = pool()
    async with p.acquire() as conn:
        row = await conn.fetchrow(
            "DELETE FROM agents WHERE id = $1 RETURNING id", id
        )
        if not row:
            raise HTTPException(status_code=404, detail={"error": "Agent not found"})
    return Response(status_code=204)


@router.get("/agents/{id}/neighborhood")
async def agent_neighborhood(id: int) -> dict:
    p = pool()
    async with p.acquire() as conn:
        agent = await conn.fetchrow("SELECT * FROM agents WHERE id = $1", id)
        if not agent:
            raise HTTPException(status_code=404, detail={"error": "Agent not found"})

        outgoing = await conn.fetch(
            "SELECT * FROM influences WHERE source_agent_id = $1", id
        )
        incoming = await conn.fetch(
            "SELECT * FROM influences WHERE target_agent_id = $1", id
        )

        neighbor_ids = set()
        for inf in outgoing:
            neighbor_ids.add(inf["target_agent_id"])
        for inf in incoming:
            neighbor_ids.add(inf["source_agent_id"])

        connections = []
        for nid in neighbor_ids:
            neighbor = await conn.fetchrow("SELECT * FROM agents WHERE id = $1", nid)
            if neighbor:
                out_inf = next(
                    (i for i in outgoing if i["target_agent_id"] == nid), None
                )
                in_inf = next(
                    (i for i in incoming if i["source_agent_id"] == nid), None
                )
                w = 0.0
                direction = "incoming"
                if out_inf:
                    w = float(out_inf["weight"])
                    direction = "outgoing"
                elif in_inf:
                    w = float(in_inf["weight"])
                connections.append(
                    {
                        "agent": agent_row(neighbor),
                        "influenceWeight": w,
                        "direction": direction,
                    }
                )

        posts = await conn.fetch(
            "SELECT * FROM posts WHERE agent_id = $1 ORDER BY created_at DESC LIMIT 20",
            id,
        )
        posts_out = [
            post_row(pr, agent_name=agent["name"], agent=agent) for pr in posts
        ]

    return {
        "agent": agent_row(agent),
        "connections": connections,
        "posts": posts_out,
    }


@router.post("/influences", status_code=201)
async def create_influence(body: dict) -> dict:
    for k in ("sourceAgentId", "targetAgentId", "weight"):
        if k not in body:
            raise HTTPException(status_code=400, detail={"error": f"missing {k}"})
    p = pool()
    async with p.acquire() as conn:
        row = await conn.fetchrow(
            """INSERT INTO influences (source_agent_id, target_agent_id, weight)
               VALUES ($1, $2, $3) RETURNING *""",
            int(body["sourceAgentId"]),
            int(body["targetAgentId"]),
            float(body["weight"]),
        )
    out = influence_row(row)
    asyncio.create_task(
        neo4j_service.sync_influence_to_graph(
            int(body["sourceAgentId"]),
            int(body["targetAgentId"]),
            float(body["weight"]),
        )
    )
    return out
